ci/Jenkinsfile

```diff
@@ -0,0 +1,116 @@
+
+pipeline {
+  agent any
+  options {
+    buildDiscarder(logRotator(numToKeepStr: '10'))
+    ansiColor('xterm')
+  }
+  stages {
+    stage("Style Checks") {
+      steps {
+        sh "env"
+      }
+    }
+    stage("C++ Matrix Build") {
+      environment {
+        BUILD_MODE = 'pull-request'
+        PARALLEL_LEVEL = '4'
+        CUDA="11.5"
+      }
+      steps {
+        script {
+          parallel generateStage("cuda_build", {
+            checkout scm
+
+            withCredentials([[
+                $class: 'AmazonWebServicesCredentialsBinding',
+                credentialsId: "aws-s3-gpuci",
+                accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+            ]]) {
+              sh "./ci/build_cpp.sh"
+            }
+          })
+        }
+      }
+    }
+
+    stage("Python Matrix Build") {
+      environment {
+        CUDA="11.5"
+      }
+      steps {
+        script {
+          parallel generateStage("python_build", {
+            cleanWs (
+              deleteDirs: true,
+              externalDelete: 'sudo rm -rf %s'
+            )
+            checkout scm
+
+            withCredentials([[
+                $class: 'AmazonWebServicesCredentialsBinding',
+                credentialsId: "aws-s3-gpuci",
+                accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+            ]]) {
+              sh "./ci/build_py.sh"
+            }
+          })
+        }
+      }
+    }
+
+    stage("PR/Branch Test Stage") {
+      when { not { triggeredBy 'TimerTrigger' } }
+      environment {
+        CUDA="11.5"
+        UPLOAD_PKGS = """${sh(
+              returnStdout: true,
+              script: 'echo "$GIT_BRANCH" | grep -iq "pr" && echo "false" || echo "true"'
+          )}"""
+      }
+      steps {
+        sh "env"
+        echo "${currentBuild.buildCauses}"
+        script {
+          parallel generateStage("branch_pr_test",{
+            cleanWs (
+              deleteDirs: true,
+              externalDelete: 'sudo rm -rf %s'
+            )
+            checkout scm
+            withCredentials([[
+                $class: 'AmazonWebServicesCredentialsBinding',
+                credentialsId: "aws-s3-gpuci",
+                accessKeyVariable: 'AWS_ACCESS_KEY_ID',
+                secretKeyVariable: 'AWS_SECRET_ACCESS_KEY'
+            ]]) {
+              sh "./ci/gpu_test.sh"
+            }
+          })
+        }
+      }
+    }
+
+    stage("Nightly Test Stage") {
+      when { triggeredBy 'TimerTrigger' }
+      environment { UPLOAD_PKGS = "true" }
+      steps {
+        sh "env"
+        echo "${currentBuild.buildCauses}"
+        script {
+          parallel generateStage("nightly_test", {
+            cleanWs (
+              deleteDirs: true,
+              externalDelete: 'sudo rm -rf %s'
+            )
+            checkout scm
+            echo "Upload packages: \${UPLOAD_PKGS}"
```

Comment:
In this block, we can move the `cleanWs` and `checkout scm` steps to the shared library. This parallel block should ultimately be the same as the one above it (i.e. just `sh "./ci/gpu_test.sh"`)